### PR TITLE
data type examples

### PR DESCRIPTION
**Pull request by freddrake**
- introduce Example object for RAML 1.0
- use separate test files for 0.8 and 1.0; they will not have anything
  in common

### CODE REVIEW

File: ramlfications/utils/types.py [hunk 1]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
             data.pop("properties")
         except KeyError:
             pass
+
+    data.update(parse_examples(root.raml_version, data))
+
     return data_type_cls(**data)
+
+
+def parse_examples(raml_version, data):
```
Comment (econchick) [inline ramlfications/utils/types.py]: Similarly with the Example model, I think this would be better in `ramlfications/parser/parameters.py`

File: ramlfications/utils/types.py [hunk 2]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+    #
+    # This is the only place that knows when to supply examples and when
+    # not, and that checks when both are supplied.
+    #
+    # This is also responsible for generating structured Example objects
+    # when appropriate (RAML >= 1.0).
+    #
+    data = data or {}
+    kw = {}
+    example = data.get("example", None)
```
Comment (econchick) [inline ramlfications/utils/types.py]: The `None` is not needed - already defaults to `None` if it can't find it/is not set.

File: ramlfications/utils/types.py [hunk 3]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+
+    if raml_version == "0.8":
+        if "examples" in data:
+            del data["examples"]
+            # Should emit a lint warning if there's an examples node and
+            # we're processing RAML 0.8: authors may be expecting it to
+            # be honored.
+        kw["example"] = example
+
+    if raml_version >= "1.0":
```
Comment (econchick) [inline ramlfications/utils/types.py]: This `>=` won't work since you're comparing a string.

File: ramlfications/utils/types.py [hunk 4]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+        if "example" in data:
+            kw["example"] = parse_example(None, example)
+
+        if "examples" in data:
+            if "example" in data:
+                raise RuntimeError("example and examples cannot co-exist")
```
Comment (econchick) [inline ramlfications/utils/types.py]: Can you add a TODO comment that this should be a part of validation?

File: ramlfications/utils/types.py [hunk 5]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+            examples = data["examples"]
+            # Must be a map:
+            if not isinstance(examples, dict):
+                # Need to decide what exception to make this.
+                raise UnknownDataTypeError
```
Comment (econchick) [inline ramlfications/utils/types.py]: This isn't the appropriate exception, but can you go ahead create one in `errors.py`? Perhaps something like `InvalidValueType`, and with a helpful error message.

File: ramlfications/utils/types.py [hunk 6]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+            kw["examples"] = NodeList([parse_example(nm, val)
+                                       for nm, val in iteritems(examples)])
+        else:
+            kw["examples"] = None
```
Comment (econchick) [inline ramlfications/utils/types.py]: I think it'd be cleaner to move this to right before the if statement (I generally don't like `else` statements if we can avoid them):

```python
kw["examples"] = None
if "examples" in data:
   # do the things
```

File: ramlfications/utils/types.py [hunk 7]
```diff
@@ -63,4 +66,64 @@ def parse_type(name, raw, root):
...
+
+    return kw
+
+
+def parse_example(name, node):
+    data = dict(name=name, value=node)
+    if name:
+        data["display_name"] = BaseContent(name)
```
Comment (econchick) [inline ramlfications/utils/types.py]: `BaseContent` is meant only for `root.documentation` and for descriptions (basically wherever there could be markdown). I've just set `name` to `display_name` if `displayName` isn't set.

File: setup.py [hunk 8]
```diff
@@ -112,7 +112,7 @@ def run_tests(self):
         "all": ["requests[security]"]
     },
     tests_require=[
-        "pytest", "mock"
+        "pytest", "mock", "pytest-mock", "pytest-localserver"
```
Comment (econchick) [inline setup.py]: This could probably be separated out in its own commit.

File: tests/data/raml_08/data-type-examples.raml [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#%RAML 0.8
```
Comment (econchick) [inline tests/data/raml_08/data-type-examples.raml]: again, the name `data-type-*` is meant for actual Data Types. This can probably just be named `examples.raml`.

File: tests/data/raml_10/data-type-examples.raml [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+  with_example_structured:
+    type: object
+    properties:
+      key:
+        type: string
+        example:
+          description: Typical value.
+          value: This is a value.
+    example:
+      description: This is a typical structure.
```
Comment (econchick) [inline tests/data/raml_10/data-type-examples.raml]: Don't these keys need to map to keys defined in `properties?` Looking at [this example](https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/#example-of-how-to-define-exampleexamples-in-raml) in the spec:

```raml
properties:
  name: string
  lastname: string
example:
  name: Bob
  lastname: Marley
```

File: tests/data/raml_10/data-type-examples.raml [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    type: string
+    examples:
+      simple: "abc"
+      fancy: "two words"
+      excessive:
+        displayName: "Serious Overkill"
+        description: |
+          There's way too much text here.
+        value:
+          error: This should not be a map.
```
Comment (econchick) [inline tests/data/raml_10/data-type-examples.raml]: I thought it could be a map ([spec](https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/#single-example): 

```raml
(pii): true
strict: false
value:
  title: Attention needed
  body: You have been added to group 274
```

File: tests/data/raml_10/something.json [hunk 12]
```diff
@@ -0,0 +1,13 @@
+{"type": "object",
```
Comment (econchick) [inline tests/data/raml_10/something.json]: Could I get a better naming of the file?

File: tests/integration/data_types/test_examples_08.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def test_simple_example_structured(api):
+    #
+    # If an example is presented as a YAML map with a value key, that's
+    # just part of the value, since RAML 0.8 doesn't have any notion of
+    # annotations or other facets of an example other than the value.
+    #
+    fp = get_form_param(api, "/with_example_structured")
+    assert fp.example == {"value": "This whole map is a value."}
```
Comment (econchick) [inline tests/integration/data_types/test_examples_08.py]: I would also assert that `examples` doesn't exist.

File: tests/integration/data_types/test_examples_08.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def test_header_with_example(api):
+    h = api.resources.filter_by(name="/with_header")[0].headers[0]
+    assert h.example is True
+
+
+def test_base_uri_param(api):
+    p = api.base_uri_params[0]
+    assert p.description.raw == "account name"
```
Comment (econchick) [inline tests/integration/data_types/test_examples_08.py]: I don't think you need to test other parts that are not a part of the example.

File: tests/integration/data_types/test_examples_08.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import, division, print_function
+"""\
```
Comment (econchick) [inline tests/integration/data_types/test_examples_08.py]: New line `\` not needed

File: tests/integration/data_types/test_examples_10.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import, division, print_function
+"""\
```
Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: New line `\` not needed

File: tests/integration/data_types/test_examples_10.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    assert ex.description.raw == "No one expects the ..."
+    assert ex.display_name.raw == "Surprise Visit"
+    assert ex.strict is True
+    assert ex.value == "Spanish Inqusition!"
+
+    ex, = h.examples.filter_by(name="broken")
+    assert ex.description.raw == "Send this for a 500"
+    assert ex.display_name.raw == "DON'T DO THIS"
+    assert ex.strict is False
+    assert ex.value == "breakfast"
```
Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: Can you add a test that `example` and `examples` return a runtime error?

File: tests/data/raml_10/data-type-examples.raml (hunk 10 above)
Comment (freddrake) [inline tests/data/raml_10/data-type-examples.raml]: The example value is contained by the ``value`` key; the others are metadata for the example.

File: tests/data/raml_10/data-type-examples.raml (hunk 11 above)
Comment (freddrake) [inline tests/data/raml_10/data-type-examples.raml]: The specific example value should not have been a map, because the parameter type is specified as ``string``.  I'll adjust the example for clarity.

File: ramlfications/models/data_types.py [hunk 18]
```diff
@@ -107,6 +107,23 @@ class RAMLDataType(object):
 
 
 @attr.s
+class Example(object):
```
Comment (freddrake) [inline ramlfications/models/data_types.py]: I wonder if there should be a ``ramlfications.models.examples`` module for this, then: data types can also have examples, not only parameters.

File: ramlfications/utils/types.py (hunk 1 above)
Comment (freddrake) [inline ramlfications/utils/types.py]: Same response applies here as well, too,

File: setup.py (hunk 8 above)
Comment (freddrake) [inline setup.py]: See PR https://github.com/spotify/ramlfications/pull/144; I'll rebase this if you land that separately.

File: ramlfications/models/data_types.py [hunk 19]
```diff
@@ -114,13 +114,13 @@ class Example(object):
...
-    value        = attr.ib()
+    value        = attr.ib(repr=False)
     name         = attr.ib(default=None)
     description  = attr.ib(default=None, repr=False)
-    display_name = attr.ib(default=None)
+    display_name = attr.ib(default=None, repr=False)
 
     # Not sure when validation of examples should get done; leave for now.
-    strict       = attr.ib(default=True)
+    strict       = attr.ib(default=True, repr=False)
```
Comment (econchick) [inline ramlfications/models/data_types.py]: Did you see my [comment](https://github.com/spotify/ramlfications/pull/143#discussion_r118041317) about `annotation_name`?

File: ramlfications/utils/types.py [hunk 20]
```diff
@@ -82,48 +84,52 @@ def parse_examples(raml_version, data):
...
         if "example" in data:
             kw["example"] = parse_example(None, example)
 
+        kw["examples"] = None
         if "examples" in data:
             if "example" in data:
-                raise RuntimeError("example and examples cannot co-exist")
+                # TODO: Emit a lint warning during validation.
+                raise InvalidRAMLStructureError(
+                    "example and examples cannot co-exist")
```
Comment (econchick) [inline ramlfications/utils/types.py]: Nitpick: can you add single quotes around the properties for readability? e.g. `"'example' and 'examples' cannot co-exist"`

File: ramlfications/utils/types.py [hunk 21]
```diff
@@ -82,48 +84,52 @@ def parse_examples(raml_version, data):
...
             examples = data["examples"]
             # Must be a map:
             if not isinstance(examples, dict):
                 # Need to decide what exception to make this.
-                raise UnknownDataTypeError
+                raise InvalidRAMLStructureError("examples must be a map node")
```
Comment (econchick) [inline ramlfications/utils/types.py]: Same here, add `'` around `examples`

File: ramlfications/utils/types.py [hunk 22]
```diff
@@ -82,48 +84,52 @@ def parse_examples(raml_version, data):
...
+            data["value"] = node["value"]
             if "description" in node:
-                data["description"] = BaseContent(data["description"])
-            if "displayName" in data:
-                data["display_name"] = BaseContent(data.pop("displayName"))
+                data["description"] = BaseContent(node["description"])
+            if "displayName" in node:
+                data["display_name"] = node["displayName"]
+            if "strict" in node:
+                data["strict"] = node["strict"]
```
Comment (econchick) [inline ramlfications/utils/types.py]: I think with these things (side-stepping the eventual adding of `NOTHING` or some type of `Undefined` type), I think it'd be cleaner if it was like this:

```python
data["description"] = BaseContent(node.get("description", ""))
data["display_name"] = node.get("displayName", name)
data["strict"] = node.get("strict", True)
```

You'd be able to remove the following 2 lines, too.

File: tests/integration/data_types/test_examples_10.py [hunk 23]
```diff
@@ -98,24 +102,46 @@ def test_header_with_mutiple_examples(api):
...
+
+def test_disallows_example_with_examples():
+    ramlfile = os.path.join(RAML_10, "broken-example-with-examples.raml")
+    loaded_raml = load_file(ramlfile)
+    conffile = os.path.join(RAML_10, "test-config.ini")
+    config = setup_config(conffile)
+    try:
+        parse_raml(loaded_raml, config)
+    except errors.InvalidRAMLStructureError as e:
+        assert str(e) == 'example and examples cannot co-exist'
```
Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: The way to assert that an exception has been raised is not with a try/except but with a `with pytest.raises`:

```python
with pytest.raises(errors.InvalidRAMLStructureError) as e:
    parse_raml(loaded_raml, config)

msg = # error message
assert msg in e.value.args[0]
```

Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: and same with the example below.

File: ramlfications/utils/types.py [hunk 24]
```diff
@@ -119,17 +119,10 @@ def parse_example(name, node):
...
-            data["value"] = node["value"]
-            if "description" in node:
-                data["description"] = BaseContent(node["description"])
-            if "displayName" in node:
-                data["display_name"] = node["displayName"]
-            if "strict" in node:
-                data["strict"] = node["strict"]
-    if "display_name" not in data:
-        data["display_name"] = name
+    if isinstance(node, dict) and "value" in node:
```
Comment (econchick) [inline ramlfications/utils/types.py]: couldn't you also remove the `and "value" in node:` and do similarly `data["value"] = node.get("value")?  I know there _has_ to be a value, but that can be caught in validation.

File: tests/integration/data_types/test_examples_10.py [hunk 25]
```diff
@@ -130,18 +130,16 @@ def test_disallows_example_with_examples():
     loaded_raml = load_file(ramlfile)
     conffile = os.path.join(RAML_10, "test-config.ini")
     config = setup_config(conffile)
-    try:
+    with pytest.raises(errors.InvalidRAMLStructureError) as e:
         parse_raml(loaded_raml, config)
-    except errors.InvalidRAMLStructureError as e:
-        assert str(e) == 'example and examples cannot co-exist'
+    assert str(e.value) == 'example and examples cannot co-exist'
```
Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: double quotes please

File: tests/integration/data_types/test_examples_10.py [hunk 26]
```diff
@@ -130,18 +130,16 @@ def test_disallows_example_with_examples():
...
     ramlfile = os.path.join(RAML_10, "broken-non-map-examples.raml")
     loaded_raml = load_file(ramlfile)
     conffile = os.path.join(RAML_10, "test-config.ini")
     config = setup_config(conffile)
-    try:
+    with pytest.raises(errors.InvalidRAMLStructureError) as e:
         parse_raml(loaded_raml, config)
-    except errors.InvalidRAMLStructureError as e:
-        assert str(e) == 'examples must be a map node'
+    assert str(e.value) == 'examples must be a map node'
```
Comment (econchick) [inline tests/integration/data_types/test_examples_10.py]: same here, double quotes

File: ramlfications/utils/types.py (hunk 24 above)
Comment (freddrake) [inline ramlfications/utils/types.py]: Not in this case.  When there's a ``value`` key, the entire interpretation changes.  If there isn't a value key, we *already* have the value, and we don't want to clobber it.  Only when we have the value key can the other bits of metadata be present.

**Comment by freddrake** [comment]
Ready for review again.

File: ramlfications/utils/types.py (hunk 24 above)
Comment (econchick) [inline ramlfications/utils/types.py]: ahh okay :+1:

**Comment by freddrake** [comment]
Commits have been squashed.

**Comment by sichvoge** [comment]
Awesome! :)
